tools/nnyeah/nnyeah/Reworker.cs

```diff
@@ -0,0 +1,488 @@
+using System;
+using System.IO;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using Mono.Cecil.Rocks;
+using System.Linq;
+using System.Collections.Generic;
+
+#nullable enable
+
+namespace nnyeah {
+	public class Reworker {
+		Stream stm;
+		ModuleDefinition? module;
+		TypeDefinition? embeddedAttributeTypeDef;
+		TypeDefinition? nativeIntegerAttributeTypeDef;
+		TypeReference? nativeIntegerAttributeTypeRef;
+		TypeReference? compilerGeneratedAttributeTypeRef;
+		TypeReference? embeddedAttributeTypeRef;
+		TypeReference? nintTypeReference;
+		TypeReference? nuintTypeReference;
+
+		Dictionary<string, Transformation>? methodSubs;
+		Dictionary<string, Transformation>? fieldSubs;
+
+		public Reworker (Stream stm)
+		{
+			this.stm = stm;
+		}
+
+		public void Load ()
+		{
+			if (module is not null)
+				return;
+			module = ModuleDefinition.ReadModule (stm);
+		}
+
+		public bool NeedsReworking ()
+		{
+			CheckModule ();
+
+			// simple predicate for seeing if there any references
+			// to types we need to care about.
+			// IntPtr is for future handling of types that
+			// descend from NObject and need to have the constructor
+			// changed to NHandle
+			return module!.GetTypeReferences ().Any (
+				tr =>
+					tr.FullName == "System.nint" ||
+					tr.FullName == "System.nuint" ||
+					tr.FullName == "System.nfloat" ||
+					tr.FullName == "System.IntPtr"
+				);
+		}
+
+		void CheckModule ()
+		{
+			if (module is null)
+				throw new Exception ("Module is not loaded. Call Load first.");
+		}
+
+		public void Rework (Stream stm)
+		{
+			CheckModule ();
+
+			// get the types that we need to refer to later
+			FetchCompilerGeneratedAttributeTypeRef ();
+			AddEmbeddedAttributeIfNeeded ();
+			AddNativeIntegerAttributeIfNeeded ();
+			module!.TryGetTypeReference ("System.nint", out nintTypeReference);
+			module!.TryGetTypeReference ("System.nuint", out nuintTypeReference);
+
+			// load the substitutions
+			methodSubs = LoadMethodSubs ();
+			fieldSubs = LoadFieldSubs ();
+
+			foreach (var type in module!.Types) {
+				ReworkType (type);
+			}
+
+			module!.Write (stm);
+			stm.Flush ();
+		}
+
+		void ReworkType (TypeDefinition definition)
+		{
+			foreach (var field in definition.Fields) {
+				ReworkField (field);
+			}
+			foreach (var property in definition.Properties) {
+				ReworkProperty (property);
+			}
+			ReworkMethods (definition.Methods);
+			foreach (var @event in definition.Events) {
+				ReworkEvent (@event);
+			}
+			foreach (var innerType in definition.NestedTypes) {
+				ReworkType (innerType);
+			}
+		}
+
+		void ReworkMethods (IEnumerable<MethodDefinition> methods)
+		{
+			foreach (var method in methods) {
+				ReworkMethod (method);
+			}
+		}
+
+		void ReworkMethod (MethodDefinition method)
+		{
+			var nativeTypes = new List<bool> ();
```

Comment:
I'm confused a bit on the `nativeTypes` state that is being cleared in multiple places. How about just new'ing the list in the loop or having ReworkTypeReference instance one?

Reply:
Possible. I don't like newing up too many objects that don't need to be, especially in a loop.

Reply:
With regards to parameterizing - yeah, but a lot of them differ subtle ways.
There are two flavors of `CompareTo` that have different parameters x 2 instance types.
There are two flavors of `TryParse`...
There are four flavors of `ToString`...
There are four flavors of `Parse`...

So yeah, I could write a function that takes the instance type and the parameter list and the call site is not going to look appreciably different.

I mean, I can see it looking like this:
```
mref = ToStringMref (module.TypeSystem.IntPtr, EmptyToStringParameterList());
```
Which cleans up that code a little, but I still need to pass in a unique set of instructions to the `Transformation` ctor and these don't parameterize well because some of them are different in one place, some different in two places, etc. It makes the transformation factory hard to read and makes the calling code harder to understand ("what am I building here?").
I know it feels like a lot of repetition inline, but personally I read each one like its own paragraph that explains what it does rather than hiding it elsewhere. Because of the uniqueness of each block, I feel it detracts from the readability and maintainability.